grails-app/services/org/pih/warehouse/inventory/InventorySnapshotService.groovy

```diff
@@ -464,155 +464,207 @@ class InventorySnapshotService {
-                    "Name"          : product.displayNameWithLocaleCode,
-                    "Product Family": product?.productFamily?.name ?: '',
-                    "Category"      : product.category.name,
-                    "Formulary"     : product.productCatalogsToString(),
-                    "Tag"           : product.tagsToString(),
-                    "Unit Cost"     : product.pricePerUnit ?: ''
-            ]
-            row.put("Opening", balanceOpening)
-            def includeRow = balanceOpening || balanceClosing || quantityAdjustments
-            transactionTypes.each { TransactionType transactionType ->
-                String translatedTransactionTypeName = transactionType?.name
-                if (transactionType?.name?.contains(Constants.LOCALIZED_STRING_SEPARATOR)) {
-                    translatedTransactionTypeName = LocalizationUtil.getLocalizedString(transactionType?.name)
+                if (orderBy) {
+                    order(orderBy, 'desc')
                 }
-                def quantity =
-                        transactionData.find {
-                            it.productCode == product.productCode && transactionType.compareName(it?.transactionTypeName)
-                        }?.quantity?:0
-                row[translatedTransactionTypeName] = quantity
-                includeRow = quantity != 0 ? true : includeRow
             }
+        } as List<TransactionEntry>
+    }
 
-            row.put("Adjustments", quantityAdjustments)
-            row.put("Closing", balanceClosing)
-
-            if (includeRow) {
-                data << row
-            }
+    Map<Product, Map<String, Integer>> getDetailedTransactionReportData(Map<Product, List<TransactionEntry>> transactionEntries) {
+        return transactionEntries.collectEntries { product, entriesForProduct ->
+            Map<String, Integer> totalsByType = entriesForProduct
+                    .groupBy { entry ->
+                        entry.transaction.transactionType.name
+                    }
+                    .collectEntries { transactionTypeName, entriesByType ->
+                        Integer total = entriesByType.sum { entry ->
+                            Math.abs(entry.quantity as Integer)
+                        }
+                        [(transactionTypeName): total]
+                    }
+
+            [(product): totalsByType]
         }
-        return data.sort { it."Code" }
     }
 
-    def getTransactionReportSummary(Location location, List<Category> categories, List<Tag> tags, List<ProductCatalog> catalogs, Date startDate, Date endDate) {
+    List<Object> getTransactionReport(Location location, List<Category> categories, List<Tag> tagsList, List<ProductCatalog> catalogsList, Date startDate, Date endDate, Boolean includeDetails) {
+        List<TransactionCode> adjustmentTransactionCodes = [
+                TransactionCode.CREDIT,
+                TransactionCode.DEBIT
+        ]
```

Comment:
What about PRODUCT_INVENTORY transactions? Baseline transactions won't have any adjustment initially, but if we have any backdated transactions, they could. For example (the table is ordered by transactionDate):
```
dateCreated | transactionDate | transactionType   | quantity
Jan 1       | Jan 1           | CREDIT            | 10
Jan 4       | Jan 2           | CREDIT            | 20 -> Was backdated
Jan 3       | Jan 3           | PRODUCT_INVENTORY | 10
```
so in that scenario, we can't simply sum up the adjustment transactions because the product inventory now has an implied adjustment of -20. The actual QoH is 10, but if we simply sum up the adjustments, we'd get QoH == 30.